src/log-store/src/kafka/util/record.rs

```diff
@@ -0,0 +1,669 @@
+            .await
+            .map_err(|e| {
+                GetClientSnafu {
+                    topic: &self.ns.topic,
+                    error: e.to_string(),
+                }
+                .build()
+            })?
+            .producer;
+
+        // Stores the offset of the last successfully produced record.
+        let mut last_offset = None;
+        for entry in self.entries {
+            for record in build_records(entry, self.max_record_size >> 1) {
+                let kafka_record: KafkaRecord = record.try_into()?;
+                // Records of a certain region cannot be produced in parallel since their order must be static.
+                let offset = producer.produce(kafka_record).await.map(Offset).context(
+                    ProduceRecordSnafu {
+                        topic: &self.ns.topic,
+                    },
+                )?;
+                last_offset = Some(offset);
+            }
+        }
+        // Safety: there must be at least one record produced when the entries are guaranteed not empty.
+        Ok(last_offset.unwrap())
+    }
+}
+
+fn check_records(records: &[Record]) -> Result<()> {
+    let len = records.len();
+    ensure!(
+        len > 0,
+        IllegalSequenceSnafu {
+            error: "Empty sequence"
+        }
+    );
+
+    let mut sequence = Vec::with_capacity(len);
+    let mut entry_ids = HashSet::with_capacity(len);
+    let mut namespaces = HashSet::with_capacity(len);
+    let mut checksum_matched = HashSet::with_capacity(len);
+    for record in records {
+        sequence.push(i32::from(record.tp));
+        entry_ids.insert(record.entry_id);
+        namespaces.insert(&record.ns);
+        checksum_matched.insert(record.checksum == crc32fast::hash(&record.data));
+    }
+
+    ensure!(
+        entry_ids.len() == 1,
+        IllegalSequenceSnafu {
+            error: "Non-unique entry ids"
+        }
+    );
+
+    ensure!(
+        namespaces.len() == 1,
+        IllegalSequenceSnafu {
+            error: "Non-unique namespaces"
+        }
+    );
+
+    ensure!(
+        checksum_matched.len() == 1 && checksum_matched.contains(&true),
+        ChecksumMismatchedSnafu
+    );
+
+    if len == 1 {
+        // A sequence with a single record must contain only a Full record.
+        if sequence[0] != i32::from(RecordType::Full) {
+            return IllegalSequenceSnafu {
+                error: "Missing Full record",
+            }
+            .fail();
+        }
+    } else {
+        // A sequence with multiple records must start with a First record and end with a Last record.
+        // In addition, the Middle records must in order.
+        let prefix = (0..(len - 1) as i32).collect::<Vec<_>>();
+        ensure!(
+            sequence[0..len - 1] == prefix
+                && sequence.last().unwrap() == &i32::from(RecordType::Last),
+            IllegalSequenceSnafu {
+                error: "Out of order",
+            }
+        );
```

Comment:
```suggestion
        ensure!(
            sequence[..len - 1]
                .windows(2)
                .all(|seq| seq[0] + 1 == seq[1])
                && sequence[len - 1] == i32::from(RecordType::Last),
            IllegalSequenceSnafu {
                error: "Out of order",
            }
        );
```

Reply:
Checking records is now performed while consuming records.